### PR TITLE
yass-cli 1.14.0 (new formula)

### PR DESCRIPTION
**Pull request by Chilledheart**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/y/yass.rb
Comment (bevanjkay) [inline Formula/y/yass.rb:1]: Because there's already a cask for `yass` I wonder if calling this `yass-cli` would help with the separation.

Comment (Chilledheart) [reply Formula/y/yass.rb:1]: That sounds reasonable.

Comment (bevanjkay) [reply Formula/y/yass.rb:1]: Because the cask also installs the `yass` binary, we would either need to make the formula and cask conflict, or remove the binary from the cask.

Comment (Chilledheart) [reply Formula/y/yass.rb]: See https://github.com/Homebrew/homebrew-cask/pull/184675

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:12]: Do we need this? Does `brew livecheck yass-cli` return wrong results if we remove it?

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:16]: Put this inside an `on_macos` block.

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:28]:
```suggestion
```
This is implied by `uses_from_macos "zlib"`.

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:34]:
```suggestion
      -DBUILD_SHARED_LIBS=on
```

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:50]:
```suggestion
    system "cmake", "-S", ".", "-B", "build", "-G", "Ninja", *args, *std_cmake_args
```
Pass `std_cmake_args` last.

Do we really need to use `ninja`?

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:53]: Why aren't we using the install target?

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:68]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

**Review by carlocab — COMMENTED** [review]
Thanks for the PR! A few comments.

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:46]:
```suggestion
      -DUSE_SYSTEM_NGHTTP2=on
      -DUSE_TCMALLOC=off
      -DUSE_MIMALLOC=off
      -DUSE_SQLITE=off
      -DUSE_SYSTEM_JSON=on
    ]
```
Let's avoid using vendored stuff unintentionally. You'll probably also need
```ruby
depends_on "nlohmann-json" => :build
```

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:34]: It is required by yass project and we're unlikely to remove it.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:53]: Install target is used for GNU-compatible system such as install systemd files and etc. it will introduce some noise on homebrew directory.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:68]: We can build unittest binary but it is unlikely to be shipped inside a bottle.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:50]: It is used only by compile-time and slightly faster than gnu make.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:50]: We don't like to adopt incompatible flags from std args in any.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:12]: I tested it when the project name is still yass. Does it change now?

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:68]: We don't need a unit test binary. From the linked documentation:

> Some advice for specific cases:
>
> - If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream’s documentation / source examples. A good example is [tinyxml2](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/t/tinyxml2.rb)’s test, which writes a small C++ source file into the test directory, compiles and links it against the tinyxml2 library and finally checks that the resulting program runs successfully.
> - If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
> - If the software cannot function without credentials or requires a virtual machine, docker instance, etc. to run, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected. This is preferred over mocking a dependency.
> - Homebrew comes with a number of [standard test fixtures](https://github.com/Homebrew/brew/tree/master/Library/Homebrew/test/support/fixtures), including numerous sample images, sounds, and documents in various formats. You can get the file path to a test fixture with e.g. test_fixtures("test.svg").

See the linked documentation for more details.

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:50]: It also requires installing `ninja` whereas `make` is built-in. Let's just use `make` unless it is substantially slower.

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:34]: Why is it required?

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:53]: That's fine. Let's just it.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:34]: For example, it compiles an upstream libc++ library and is linked statically into yass because macOS below 12.0 doesn't have the correct c++20 std library. It will cause troubles if it is linked shared library. There are more than 10 static libraries (mostly, c++ libraries) inside yass project and untested for shared linkage. We just avoid to run into dependency hell.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:68]: We can provide some configure validation feature inside cli tool just like the gui one(see homebrew-cask) later. But currently, version check is what we have now.

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:34]: But we're not building on versions of macOS below macOS 12, so we shouldn't be using a custom libc++.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:50]: I thought ninja is recommended than make because yass has more than one thousand source targets just like llvm recipe. It is fine to use just make for slightly slower compiling performance.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:34]: Okay. It breaks on macOS 13.0 when building with system libc++ because missing features of std::filesystem (I tried system libc++ at first commit).

Comment (carlocab) [reply Formula/y/yass-cli.rb:34]: Use the `libc++` from `llvm` like this:

https://github.com/Homebrew/homebrew-core/blob/5ffbd6b840b1a95da482f8b2a10729e76628d749/Formula/d/dwarfs.rb#L80-L86

You'll need to drop `=> :build` from the `llvm` dependency.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:34]: It is too heavy to depends on versioned llvm even if appleclang is equal or more than 1500. At least we use the upstream libc++ which is the same revision used in production branch of chrome and chromium which is well tested on macOS. 

We don't get the same guarantee from versioned release of llvm.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:12]:
```
➜  ~ brew livecheck yass-cli --debug
/opt/homebrew/Library/Homebrew/brew.rb (Formulary::FromNameLoader): loading yass-cli
/opt/homebrew/Library/Homebrew/brew.rb (Cask::CaskLoader::NullLoader): loading yass-cli

Formula:          yass-cli
Livecheckable?:   Yes

URL (stable):     https://github.com/Chilledheart/yass/releases/download/1.13.1/yass-1.13.1.tar.bz2
Strategy:         GithubLatest
/usr/bin/env PATH=/opt/homebrew/opt/gh/bin:/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/bin:/opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin:/opt/homebrew/Library/Homebrew/shims/shared:/usr/bin:/bin:/usr/sbin:/sbin HOME=/Users/hky gh auth token --hostname github.com
/usr/bin/env PATH=/opt/homebrew/opt/gh/bin:/opt/homebrew/Library/Homebrew/vendor/bundle/ruby/3.3.0/bin:/opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin:/opt/homebrew/Library/Homebrew/shims/shared:/usr/bin:/bin:/usr/sbin:/sbin HOME=/Users/hky gh auth token --hostname github.com
/usr/bin/env /opt/homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/4.3.19-103-g170a7ea\ \(Macintosh\;\ arm64\ Mac\ OS\ X\ 14.6.1\)\ curl/8.7.1 --header Accept-Language:\ en --retry 3 --location https://api.github.com/repos/Chilledheart/yass/releases/latest --header Accept:\ application/vnd.github+json --write-out '
'\%\{http_code\} --header Authorization:\ token\ ****** --header X-GitHub-Api-Version:2022-11-28 --dump-header /private/tmp/github_api_headers20240908-45359-g9syns
URL (strategy):   https://api.github.com/repos/Chilledheart/yass/releases/latest
Regex (strategy): /v?(\d+(?:\.\d+)+)/i

Matched Versions:
1.13.1

yass-cli: 1.13.1 ==> 1.13.1
```

It works well.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:28]: nice design!

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:16]:
```
  * line 22, col 3: `depends_on` (line 22) should be put before `on_macos` (line 18)
```

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:16]: Done.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:66]: Tracked in https://github.com/Chilledheart/yass/issues/1085

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:44]: tcmalloc and mimalloc are not available on macos (at least no working with yass).

I will add `-DUSE_SYSTEM_JSON=on`

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:44]: 截屏2024-09-08 20 00 11

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:33]: It's still gated by `if DevelopmentTools.clang_build_version < 1500` so it won't be a dependency on newer versions of macOS. The libc++ from the versioned release of llvm is also well-tested on macOS in our CI and by our users.

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:12]: I'm not sure you understood me -- does `brew livecheck` return the wrong results _if this block is removed_?

I know it works if we have this `livecheck` block, but we shouldn't use this block if it's not needed because GitHub has stricter rate limits for the API calls made by the `:github_latest` strategy.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: I still don't think additional 1.8G disk usage is a good option.
截屏2024-09-08 20 19 02

yass's bundled libc++ is well tested as well and we don't use llvm on macOS 13.0 because the appleclang is 1500.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: 截屏2024-09-08 20 21 57

screenshot from https://ci.chromium.org/p/chromium/g/main/console.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: 截屏2024-09-08 20 24 09

截屏2024-09-08 20 26 08

Comment (carlocab) [reply Formula/y/yass-cli.rb:33]: Disk space is cheap, and using a custom libc++ goes against https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: > we don't use llvm on macOS 13.0 because the appleclang is 1500 there.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: system libc++ is cheap if it doesn't implement the std ts. Even it is possible to like static libc++ library in llvm recipe. We are unlikely to add custom libc++ support (for yass, llvm recipe's builtin libc++ library is custom one)

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: 截屏2024-09-08 20 41 04

For macOS 13.0 build, there is no llvm packages. I don't think it is a good way to keep use system libc++ in newer macOS and use custom libc++ for  old macOS. It is burden to homebrew rules as you described as well.


https://github.com/Homebrew/homebrew-core/actions/runs/10759695738/job/29836767807?pr=183910

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:68]: We don't need a validation feature. We should just run a command some user might run using `yas_cli`.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:33]: And we always find bugs in old libc++ which is not fixed and a burden to make workaround in yass side.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:68]: yes, then we should be able to run unittests just like Gentoo-like/FreeBSD ports-like system. It is much safer and give us more guarentee.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:68]: Gentoo-like: https://github.com/microcai/gentoo-zh/blob/master/net-proxy/yass/yass-1.13.1.ebuild
FreeBSD ports like: https://bugs.freebsd.org/bugzilla/attachment.cgi?id=251559&action=edit

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb]: Yes, I saw that, but I don't want to rely on those `cmake_dependent_option` declarations not changing.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb]: `cmake_dependent_option` means the if prerequirement condition is not met, there is no such cmake option. It makes no effect even if you pass an ON to such option.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:12]: Yes, the output is different somehow.

截屏2024-09-08 20 53 33

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:12]: Please notice two things: 
1. The livecheckable field is now "No".
2. the version becomes an integer instead of point-release style.

**Comment by Chilledheart** [comment]
Added `USE_TCMALLOC` argument and relative checks for linux variant.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb]: We can add it for linux through.

File: Formula/y/yass-cli.rb
Comment (carlocab) [inline Formula/y/yass-cli.rb:11]:
```suggestion
    regex(/^v?(\d+(?:\.\d+)+)$/i)
```
If the problem is picking up weird versions, then we should use a `regex` instead of `:github_latest`.

File: Formula/y/yass-cli.rb
Comment (carlocab) [reply Formula/y/yass-cli.rb:37]: > For macOS 13.0 build, there is no llvm packages.

I don't know what this means.

> I don't think it is a good way to keep use system libc++ in newer macOS and use custom libc++ for old macOS.

I disagree. We should use system libc++ where we can, and LLVM libc++ where we can't.

> It is burden to homebrew rules as you described as well.

It is not.

**Review by carlocab — CHANGES_REQUESTED** [review]
I'm not sure why you keep marking unaddressed reviews as resolved, so to be sure my feedback is addressed before merging, I'm requesting changes here.

**Comment by Chilledheart** [comment]
Feel free to comment. I thought we have settled down several issues.

File: Formula/y/yass-cli.rb
Comment (Chilledheart) [reply Formula/y/yass-cli.rb:11]: I don't see any problem we should need to resolve here.

**Comment by Chilledheart** [comment]
As I commented, we don't always use llvm packages if the appleclang(provided by xcode/CLI) is 1500 or newer such as it is on macOS 13.0. So it is impossible to pick up it when we don't need it at all. And I don't think it is right way to depend on an untesetde c++ standard library (it is different story from versioned 3rd-party library).

As a developer, I am strongly unhappy if someone tells me you cannot use the latest c++ features because we are depending on a versioned llvm package or an old system libc++ library. And that's the story of libc++ from point-released llvm package. From my memory, I am keeping fixing kinds of bugs and compiling issues with the old or new libc++.

As reference, if you develop native program in android, you're recommended to compile and use the latest libc++ library because the system one is legacy and there're no 3rd-party libc++ apk for this. Or you can compile statically to latest libc++ library in the NDK.

**Comment by Chilledheart** [comment]
As above, we are not against the rules of homebrew which keep most usage of the shared libraries. We are using the correct way to keep away from someone who tells us that YOU'RE NOT USING THE LATEST C++ FEATURES.

**Comment by Chilledheart** [comment]
Besides, I don't like libc++ developers because they makes mess a lot and don't fix it even I upload the workaround and bugs, unlike the other LLVM developers.

File: Formula/y/yass-cli.rb
Comment (bevanjkay) [reply Formula/y/yass-cli.rb:11]: That isn't the suggestion. We don't want to use the `github_latest` strategy because it uses the API and we prefer not to do that.
If you use the `regex` provided in the suggestion above it will only match versions that use your current version scheme.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:11]: It is the different from git's tags and GitHub releases where GitHub releases depend on git tags but not the vice versa.

Yass currently uses GitHub as the main git host site. It makes sense if yass releases other releases without GitHub release.

However, it is possible that some tags are pushed but releases are not ready yet. It would be convenient if fixes are push before releases are finally out.

And there are two release lines which are stable release and legacy release. Both of them use the same point release style such as x.y.z. It would be inconvenient if regex cannot tell them from git tag name. Both of these new releases are pushed regularly in the life of time. However newest GitHub release will always be stable releases.

Comment (bevanjkay) [reply Formula/y/yass-cli.rb:11]: Thank you for the explanation, using `github_latest` makes sense in this instance.

**Comment by Chilledheart** [comment]
This revision now includes another binary (built from the same source code, but not the unittest binary) so we can test basic functionality in test step.

**Comment by Chilledheart** [comment]
And I just realized one important thing when choosing standarnd c++ library. That is if we are going to link with some homebrew's c++ library (such as [jsoncpp](https://formulae.brew.sh/formula/jsoncpp#default) which is linked to system libc++, we cannot link the custom libc++ library because their libc++ symbols conflicts. If we are going to link any of homebrew's c++ libraries, it is not helpful to link other custom libc++ library than system one because the they are not linked to it. We are going to move to jsoncpp so use yass' bundled libc++ library is a must.

If you know C++, you must know the long history of C++ library dependency such as Application of MFC dll and MFC Application's dll. Don't get you messed with homebrew's rules.

TL'DR. c++ libraries from homebrew won't get linked/work properly if you choose the libc++ library from llvm recipe. nlohmann json may be an exclusion because it's head-only library.

**Comment by Chilledheart** [comment]
If you insist on using llvm recipe in this recipe, you can start work to convert https://formulae.brew.sh/formula/jsoncpp#default https://formulae.brew.sh/formula/abseil#default https://formulae.brew.sh/formula/protobuf#default  https://formulae.brew.sh/formula/google-benchmark#default https://formulae.brew.sh/formula/googletest#default to depend on llvm recipe otherwise you're only breaking your own words and waste our time @carlocab .

File: Formula/y/yass.rb
Comment (Chilledheart) [reply Formula/y/yass.rb]: I must make it clear that the binary provided by homebrew-cask's the GUI not the yass-cli. It is totally wrong that the yass binary will be replaced by yass-cli binary.

 I will close the above PR.

File: Formula/y/yass-cli.rb
Comment (SMillerDev) [inline Formula/y/yass-cli.rb:59]:
```suggestion
    system "cmake", "-S", ".", "-B", "build", *args, *std_cmake_args
```

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:59]: Rejected due to no explicit reason.

Comment (Chilledheart) [reply Formula/y/yass-cli.rb:62]: Your reason is hard-code rule. It is not mine. As I stated previous, the std_cmake_cargs messes or will mess with args and that will make failures in building or runtime. State your reason clearly, that's what I think of homebrew as an OSS project.

**Comment by Chilledheart** [comment]
As a propose, I suggest some comment on caveats block due to the lack of support of gui on linux:

```
    on_linux do
      <<~EOS
        This formula only installs the command-line utilities by default.

        Install yass.app with Flatpak:
          flatpak install --user io.github.chilledheart.yass
      EOS
    end
```

**Comment by Chilledheart** [comment]
截屏2024-09-16 16 39 12


w up d tomorrow.

**Comment by carlocab** [comment]
Given your insistence on not applying changes maintainers have requested ([and not just mine](https://github.com/Homebrew/homebrew-core/pull/183910#discussion_r1760714360)), I don't think this formula is a good fit for Homebrew/core.

If you wish, you can host this formula in your own tap. Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

Thanks for taking the time to open a PR anyway!

**Comment by Chilledheart** [comment]
Good luck wi u s andard. Later!

**Comment by Chilledheart** [comment]
take it easy, mo*key.

**Comment by carlocab** [comment]
For future reference:

I'm not responsible for the blocking, but I likely would have been as soon as I got the chance to refer the last comment to our PLC.

Calling people "monkey" is considered a slur in many parts of the world and is a violation of our [Code of Conduct](https://github.com/Homebrew/.github/blob/69e2b6fbb5ee7b2816fd6b7f98dbeb159d6f2021/CODE_OF_CONDUCT.md).

**Pull request closed without merging**
